graph/patterns/upsert.md

```diff
@@ -0,0 +1,209 @@
+# Upsert
+
+Microsoft Graph API Design Pattern
+
+*The `Upsert` pattern is a non-destructive idempotent operation using a client-provided key, that ensures that system resources can be deployed in a reliable, repeatable, and controlled way, typically used in Infrastructure as Code (IaC) scenarios.*
+
+## Problem
+
+Infrastructure as code (IaC) defines system resources and topologies in a declarative manner that allows teams to manage those resources as they would code.
+Practicing IaC helps teams deploy system resources in a reliable, repeatable, and controlled way. 
+IaC also helps automate deployment and reduces the risk of human error, especially for complex large environments. 
+Customers want to adopt IaC practices for many of the resources managed through Microsoft Graph.
+
+Most resources' creation operations in Microsoft Graph are not idempotent in nature.
+As a consequence, API consumers that want to offer IaC solutions, must create compensation layers that can mimic idempotent behavior. 
+For example, when creating a resource, the compensation layer must check whether the resource first exists, before trying to create or update the resource.
+
+Additionally, IaC code scripts or templates usually employ client-provided names (or keys) to track resources in a predictable manner, whereas [Microsoft Graph guidelines](../GuidelinesGraph.md#behavior-modeling) suggests use of `POST` to create new entities with service-generated keys.
+
+## Solution
+
+The solution is to use an `Upsert` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `Upsert` semantics with an [alternate key](./alternate-key.md):
```

Comment:
>[alternate key](./alternate-key.md) [](http://example.com/codeflow?start=67&length=35)

this pattern should apply to any client-provided key. if the primary key is server generated then the guidance should be to define an alternate key to support idempotent creates, but if there is a natural key that is meaningful to the client, then the service should use that as the key and support upserts against when doing a PATCH to the canonical URL.  We wont force that on existing APIs that use natural keys if there is a concern about backward compatibility, but new APIs with natural keys should support upsert using those natural keys.